File: 김찬우/프로그래머스 /Lv.3/배달.java
import java.util.*;

class Solution {
    
    static List<ArrayList<Node>> list;
    static int[] visited;
    
    
    public int solution(int N, int[][] road, int K) {
        int answer = 1;

        list = new ArrayList<>();
        visited = new int[N+1];
        
        for (int i = 0; i <= N; i++) {
            list.add(new ArrayList<>());
        }
        
        for (int i=0; i < road.length; i++) {
            list.get(road[i][0]).add(new Node(road[i][0],road[i][1],road[i][2]));
            list.get(road[i][1]).add(new Node(road[i][1],road[i][0],road[i][2]));
            
        }
        
        for (int i=2; i<=N; i++){
            visited[i] = Integer.MAX_VALUE;
        }
        
        Queue<Node> queue = new ArrayDeque<>();
        queue.addAll(list.get(1));
        
        while (!queue.isEmpty()) {
            Node node = queue.poll();
            
            if (visited[node.x] <= visited[node.y] + node.z) {
                continue;
            }
            
            visited[node.x] = visited[node.y] + node.z;
            queue.addAll(list.get(node.x));
        }
        
        
        for (int i=2; i<=N; i++) {
            if (visited[i] <= K) {
                answer++;
            }
        }
        

        return answer;
    }
    
    static class Node {
        
        int y;
        int x;
        int z;
        
        public Node(int y, int x, int z) {
            this.y = y;
            this.x = x;
            this.z = z;
            
            
        }
        
    }
    
    
}

File: 조준장/level2/배달.kt
import java.util.*

class Solution {
    data class Node(val position : Int, val time : Int = 0)
    fun solution(N: Int, road: Array<IntArray>, k: Int): Int {
        var answer = 0
        
      val graph = Array(N + 1) { mutableListOf<Node>() }
        road.forEach {
            val a = it.first()
            val b = it[1]
            val c = it.last()
            graph[a].add(Node(b, c))
            graph[b].add(Node(a, c))
        }

        val queue = PriorityQueue<Node>() { o1, o2 ->
            o2.time - o1.time
        }
        val visited = IntArray(N + 1) { Int.MAX_VALUE }
        visited[1] = 0
        queue.add(Node(1))
        
        while (queue.isNotEmpty()) {
            val (targetPosition, targetTime) = queue.poll()
            
            for (next in graph[targetPosition]) {
                val totalTime = next.time + targetTime
                if (totalTime <= k && totalTime < visited[next.position]) {
                    visited[next.position] = totalTime
                    queue.add(Node(next.position, totalTime))
                }
            }
        }
        
        for (v in visited) {
            if (v != Int.MAX_VALUE) answer ++
        }

        return answer
    }
}
